src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1252,5 +1245,236 @@ internal bool CanAddParsedValue(HttpHeaderParser parser)
+                {
+                    entries[count].Key = key;
+                    return ref entries[count].Value!;
+                }
+
+                return ref GrowEntriesAndAddDefault(key);
+            }
+            else if (store is null)
+            {
+                _count++;
+                entries = new HeaderEntry[InitialCapacity];
+                _headerStore = entries;
+                ref HeaderEntry firstEntry = ref MemoryMarshal.GetArrayDataReference(entries);
+                firstEntry.Key = key;
+                return ref firstEntry.Value!;
+            }
+            else
+            {
+                return ref DictionaryGetValueRefOrAddDefault(key);
+            }
+
+            ref object? GrowEntriesAndAddDefault(HeaderDescriptor key)
+            {
+                var entries = (HeaderEntry[])_headerStore!;
+                if (entries.Length == ArrayThreshold)
+                {
+                    return ref ConvertToDictionaryAndAddDefault(key);
+                }
+                else
+                {
+                    Array.Resize(ref entries, entries.Length << 1);
+                    _headerStore = entries;
+                    ref HeaderEntry firstNewEntry = ref entries[entries.Length >> 1];
+                    firstNewEntry.Key = key;
+                    return ref firstNewEntry.Value!;
+                }
+            }
+
+            ref object? ConvertToDictionaryAndAddDefault(HeaderDescriptor key)
+            {
+                var entries = (HeaderEntry[])_headerStore!;
+                var dictionary = new Dictionary<HeaderDescriptor, object>(ArrayThreshold);
+                _headerStore = dictionary;
+                foreach (HeaderEntry entry in entries)
+                {
+                    dictionary.Add(entry.Key, entry.Value);
+                }
+                Debug.Assert(dictionary.Count == _count - 1);
+                return ref CollectionsMarshal.GetValueRefOrAddDefault(dictionary, key, out s_dictionaryGetValueRefOrAddDefaultExistsDummy);
+            }
+
+            ref object? DictionaryGetValueRefOrAddDefault(HeaderDescriptor key)
+            {
+                var dictionary = (Dictionary<HeaderDescriptor, object>)_headerStore!;
+                ref object? value = ref CollectionsMarshal.GetValueRefOrAddDefault(dictionary, key, out s_dictionaryGetValueRefOrAddDefaultExistsDummy);
+                if (value is null)
+                {
+                    _count++;
+                }
+                return ref value;
+            }
+        }
+
+        private void AddEntryToStore(HeaderEntry entry)
+        {
+            Debug.Assert(!ContainsKey(entry.Key));
+
+            if (_headerStore is HeaderEntry[] entries)
+            {
+                int count = _count;
+                if ((uint)count < (uint)entries.Length)
+                {
+                    entries[count] = entry;
+                    _count++;
```

Comment:
Would this be better as `_count = count + 1;`?

Reply:
It's just [slightly worse](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgBEBLbAcwDsJcGDmFwBuGvSbkU42o2YsASgFdeQ/DBYBhCPgAOHADYwoAZRMA3YTDET5rFWo4aWASTUmIe81CtgbshIAzAzYwIJQ2GAYTKQMABIYGHrxMNgAJia4NADeNAwFDHpQHBbYGDAMEMAAVjDRAPwMAPoAFmmZZhjQMLKFRSVlFQwcai2Qqhh9hfmFxaXllcxIDMAQEIYMuM3pwkIQvNhQAJ4A4jAYAGrYhsowijAAZgDyUACC6elsT9jKhhgAUQQHEEuDYynw+GO0wK8yGlUgvEEIzG7g4QhuWmweii6OODAAvAwZLMCqMKlBDptEcjyQw3lBIscACqtWC4VobdKEhioGEMUkDBbDYgoemfAFqE7MiCmbqwAAUqQyJklGBODBgUuOAEpBXlqP1+hxHgwFW0OiY5T0RrgEpaoGqTgBtAC6mqlHBsesNRoKBr9frpEzGROaIamgsDI1NCoVynJOojDAAPGaE2odVr1V7cCwADJa7gYVo+6P9APlv3Zko2Z0R91EmvQqNV8YQSYAak7/LbTAA7L3AwBfVsFUfUMcMc5XG53B4vd6fb6PX7/BXNlgAaRgup5m+ut16gon/UFcMWTDFH3STpZsvlMFISodd496t1+qnJrNFpVXRtEF7X/O83XfWtcDLQNKyrYMO1DdtJiHINY3jRNkzTNDMxrXMCyLEsoKrGC+3A3N63gjBG3AltfRIxCEOTTsGHIZDy2IQcp36U8RxoKcZ0PecnleG8VzXDAN21bddx1fdJIE49aIYbiCnPQZL1gU1qjqRppwueSF2E5cfj+cTlU6b5cDAEo9HlBgAGtpK/RT+i0+oMCaQQbTDdp/2tWBWJ/BVPNgW1gM6UD3Rw70p2I6NHmgM06Q4Hk6FEEZU3omIADIsvStMorzQteGLVo0o4btCPLWK20ChzjhYAEAEdlBuXAJJzOsOFdKTdUqvtqro9iGA00jOu6+SAEJWOjZTywnTjCjgyYeXDCjpsKVaux7BayVQjMMCTCiMqwg6Crw4qCJ2hgBvLAryMmbqd3xIk6vWv0hpGu6GxYSa3qU3jnMDD6nmnKAIAAdydXM3l4dIROM9c6p1VjZv6GBDFwSpAuCrG7V4P5DD6wobv6Ta1G7P6Cp5XgYHBsLVW1Z00QxQwsRxMA8VdP6/06PzKibT0AiukazIZj8GEeDgoEEN8iRGgBZGB8GgY55aODkbhYGcGSZNhymwBcTC1fx2og5Grsl6XAUkp6eVeq7gdNS2ZbkucYCmqdUcKdHMZih3+2GkHOGiDgDiOM49Ldgyly+BHxKRlGAejEbXJ004wchwXcBhuGjNXEyX3/CyrI4GyEqRv3Ab9MooFGu0iUL8LGddHUeatR8Paro1ArOoqSsJIkdewFk2RsTlDHSIm/RJoGA5GnReAsEwMBlYP9kOE4c/h/PEekv6verDGYCume/SH+qHlwDgAC8YAVT6s7QOvzv7lM03Ic2u+jNuAJCgWOtsF/QMItXzagllLQQAA5WmstA6mjur3fCrQGAAD4UHMS5ldfozsMDQMzh+Hqdtdx/X6I7cBVs8F3h+m7TufYD7/UnF3FOtQ3JNAXkvKAK8IBr1DhvY4W885iUbiYYu1lbIVy7qfBgNc648mEY6ZurcfK8w7n9GRuwQ5hw1ESGmdMeFaOOCmUWUBRGl3lE/VOGAUEKnPqydk49J7c2Ue3LyDANHr3Dn9eKsAojIPkW+ZsKI65TyNFI/o7jeHhxYDeU29UnpPwPG7T+dCrrfGAMobg0TcCY04QqCJBjtBHQJGGZMcBmLJLbGQnQhhjCaKRGraWrRNb8SjkJGOokC75L4U/OqFjlAxG2F08OLSjzR23mJIEIIMBgghFCCpfp5pMJBpYpo+i+EjMEoucZBdjGmLLrXCR0YpHqL2JE7RZo1nhyMQ6PZ5iqgsOiCgpRDo+a0LbMw7S7lpFux5PPDYtT164AaRrQwWtI6jLads9cQyTg9N3H0gZOxTkGI2fcSFgiTKTNBOCSEuo/qBTKEeUK+MakhIrFgjaEYKZXXoaQueINCV3ETopCcizyQmCpFsdUyhoj0wUR+XIgpiAhF2TYEu+yGBPX5MK+5nyGDyUCIpGVxi7zyNueXeFsq3LfKPJVG6tsXrEKnPJHkjKFJcRoGyjwlIbiBwyAcQw+IIg8piKKyyYiEogAYK4JqLUMBhGMNcouYqPVQBQYKxSF5hg+PSA6/EliWiZHdWY6AirSEhDWBsBgvrWpqpDSm2uEASwmD1VOBNSbxW2TDBW0NaVAzfljDWgtYKV7HD0HfGSxSGAYDbTACAjwFTMDoDqEJUihpDoas1XNABVJE2BHiaDeLgFMQ7rFNv2TqCxxaoAsB2PmiVeApB0CfnKEoxUdD6COCCA4LBXi7CpK4PgPQsSY3mUaehPtj6SP9gwQ2sBeD+BzRjPJ+67lFvaDuvdyaN3MotdQK1FJOWYxuDAbkZBJX8HBrwYxEb01HoYJA7AGhroMG4BcNKrKgA===) as it has to store an extra value on the stack